tabular/src/autogluon/tabular/models/tabicl/tabicl_model.py

```diff
@@ -0,0 +1,174 @@
+"""
+Code Adapted from TabArena: https://github.com/autogluon/tabrepo/blob/main/tabrepo/benchmark/models/ag/tabicl/tabicl_model.py
+Model: TabICL
+Paper: TabICL: A Tabular Foundation Model for In-Context Learning on Large Data
+Authors: Jingang Qu, David Holzmüller, Gaël Varoquaux, Marine Le Morvan
+Codebase: https://github.com/soda-inria/tabicl
+License: BSD-3-Clause
+"""
+
+from __future__ import annotations
+
+import logging
+
+import pandas as pd
+
+from autogluon.common.utils.pandas_utils import get_approximate_df_mem_usage
+from autogluon.common.utils.resource_utils import ResourceManager
+from autogluon.core.models import AbstractModel
+from autogluon.tabular import __version__
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Verify if crashes when weights are not yet downloaded and fit in parallel
+class TabICLModel(AbstractModel):
+    ag_key = "TABICL"
+    ag_name = "TabICL"
+    ag_priority = 65
+
+    def get_model_cls(self):
+        from tabicl import TabICLClassifier
+
+        if self.problem_type in ["binary", "multiclass"]:
+            model_cls = TabICLClassifier
+        else:
+            raise AssertionError(f"Unsupported problem_type: {self.problem_type}")
+        return model_cls
+
+    @staticmethod
+    def _get_batch_size(n_cells: int):
+        if n_cells <= 4_000_000:
+            return 8
+        elif n_cells <= 6_000_000:
+            return 4
+        else:
+            return 2
+
+    def _fit(
+        self,
+        X: pd.DataFrame,
+        y: pd.Series,
+        num_cpus: int = 1,
+        num_gpus: int = 0,
+        **kwargs,
+    ):
+        try:
+            import tabicl
+        except ImportError as err:
+            logger.log(
+                40,
+                f"\tFailed to import tabicl! To use the TabICL model, "
+                f"do: `pip install autogluon.tabular[tabicl]=={__version__}`.",
+            )
+            raise err
+
+        from torch.cuda import is_available
+
+        device = "cuda" if num_gpus != 0 else "cpu"
```

Comment:
Will it be able to support other accelerator such as Apple Silicon given we have its native support in AutoMM now? Is it easy to test?

Reply:
We can test if we find time after merging the initial PRs. I haven't personally tried.